### PR TITLE
Fix GdipCreatePath2 validation and add tests

### PR DESCRIPTION
**Pull request by hughbe**
Fixes https://github.com/mono/libgdiplus/issues/401
Fixes https://github.com/mono/libgdiplus/issues/400

### CODE REVIEW

**Comment by hughbe** [comment]
Rebased /cc @marek-safar @filipnavara

**Comment by hughbe** [comment]
Friendly ping @marek-safar @fillipnavara - I’m back with a laptop today and have some PRS that depend on this one I fancy sending in

File: src/general.c [hunk 1]
```diff
@@ -227,28 +227,30 @@ gdip_get_display_dpi ()
 GpPointF *
 convert_points (const GpPoint *point, int count)
 {
-	int i;
+	if (!point || count < 0)
```
Comment (filipnavara) [inline src/general.c]: I'd suggest `count <= 0` since it would return NULL anyway in the next step.

File: src/general.c [hunk 2]
```diff
@@ -227,28 +227,30 @@ gdip_get_display_dpi ()
...
 
 	return retval;
 }
 
 GpRectF *
 convert_rects (const GpRect *rect, int count)
 {
-	int i;
+	if (!rect || count < 0)
+		return NULL;
```
Comment (filipnavara) [inline src/general.c]: Likewise, `count <= 0`.

File: src/graphics-path.c [hunk 3]
```diff
@@ -199,6 +199,63 @@ append_curve (GpPath *path, const GpPointF *points, GpPointF *tangents, int offs
...
+gdip_validate_path_types (const BYTE *types, INT count)
+{
+	const BYTE *currentType = types;
+	INT remaining = count;
+
+	if (count == 1)
+		return TRUE;
+
+	while (TRUE) {
+		remaining -= 1;
```
Comment (marek-safar) [inline src/graphics-path.c]: Just use `--` and `++` everywhere

File: src/general.c (hunk 1 above)
Comment (hughbe) [reply src/general.c]: Can't change this - on my PC this causes test failures in `testpathgradientbrush`

Will investigate further in another PR?

File: src/graphics-path.c (hunk 3 above)
Comment (hughbe) [reply src/graphics-path.c]: Fixed :)

File: src/general.c (hunk 1 above)
Comment (filipnavara) [reply src/general.c]: That's weird, but I am fine with postponing it.

**Pull request closed without merging**